File: src/main/java/co/fastestpath/api/schedule/TripId.java
package co.fastestpath.api.schedule;

public class TripId {

  private final String value;

  public TripId(String value) {
    this.value = value;
  }

  @Override
  public String toString() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    TripId stopId = (TripId) o;

    return value != null ? value.equals(stopId.value) : stopId.value == null;
  }

  @Override
  public int hashCode() {
    return value != null ? value.hashCode() : 0;
  }
}

File: core/src/main/kotlin/co/fastestpath/api/bootstrap/schedule/TripId.kt
package co.fastestpath.api.bootstrap.schedule

data class TripId private constructor(val value: String) {
  companion object {

    private val cache: MutableMap<String, TripId> = HashMap()

    fun of(value: String): TripId {
      var tripId = cache[value]
      if (tripId != null) {
        return tripId
      }
      tripId = TripId(value)
      cache[value] = tripId
      return tripId
    }
  }
}
